### PR TITLE
Remove `asgiref` from minimal requirements

### PR DESCRIPTION
**Pull request by Kludex**
This PR removes `asgiref` from the minimal requirements.

Changes:
- Leverage `from __future__ import annotations` to get a small diff - besides the [_collateral benefits_](https://lukasz.langa.pl/61df599c-d9d8-4938-868b-36b67fdb4448/).
- Move `asgiref` from `minimal_requirements` to `extra_requirements`.

### CODE REVIEW

**Comment by br3ndonland** [comment]
In general I think this looks fine, and I'm glad to see Uvicorn entering the `__future__`.

There's been a lot of discussion about lumping all the optional dependencies into `uvicorn[standard]` though, and asgiref isn't directly related to the other ones. Do we want to make people install all the optional dependencies just to do type checking? I'm not sure.

Related:

- encode/uvicorn#219
- encode/uvicorn#1274

**Comment by Kludex** [comment]
You have a good point. Then, are we against having it as optional but not on the standard extras? 🤔

File: uvicorn/middleware/asgi2.py [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
+
+import typing
+
+if typing.TYPE_CHECKING:
+    from asgiref.typing import (
+        ASGI2Application,
+        ASGIReceiveCallable,
+        ASGISendCallable,
+        Scope,
+    )
```
Comment (adriangb) [inline uvicorn/middleware/asgi2.py:11]: This is quite a bit of boilerplate that we'll have to have in every module

File: tests/middleware/test_proxy_headers.py [hunk 2]
```diff
@@ -1,12 +1,16 @@
-from typing import List, Union
+from __future__ import annotations
```
Comment (adriangb) [inline tests/middleware/test_proxy_headers.py:1]: Given that this `__future__` is in somewhat of a limbo, I would be reluctant to merge it.

**Review by adriangb — COMMENTED** [review]
Is this worth it? I understand that Uvicorn wants to be minimal on dependencies, but I'm not sure the internal complexity with `TYPE_CHECKING` is worth it. I also agree that due to how the extras are currently specified it's problematic to ask users to install a bunch of dependencies just to get type checking, and in fact that is contrary to the goal of not forcing dependencies on users (granted, users could look into the extra and cherry pick what they want but that's quite a bit to ask and defeats the point of extras in the first place). Somewhat tangential, but I think this would be a good fit for a `default-extras` if it existed.

**Comment by Kludex** [comment]
> Is this worth it? I understand that Uvicorn wants to be minimal on dependencies, but I'm not sure the internal complexity with TYPE_CHECKING is worth it.

This is more for `django-channel` users than to be minimal (but also): https://github.com/encode/uvicorn/issues/1305.

**Comment by Kludex** [comment]
- Let's discuss it first on https://github.com/encode/uvicorn/discussions/1527

**Pull request closed without merging**